File: smr.py
# listas descritivas de importações para certificar que o programa não é malicioso
from sys import argv # usado para carregar parametros na chamada do programa
from os import getcwd # usado para carregar o caminho do smr visualmente
from os import path # usado para filtrar o caminho vindo do getcwd

# Importações de facades
from Creator import Creator
from Deleter import Deleter
from Helper import Helper




# Mapeamento de chaves para as funções de criação e exclusão
strategy_map = {
    "create": Creator.execute,
    "delete": Deleter.execute,
    "help": Helper.execute
}

def Loader(entrada : str):
    strategy_key = entrada[0]
    key = parametros = None
    if len(entrada) > 1: 
        key = entrada[1]
    if len(entrada) > 2:
        parametros = entrada[2:]
    
    if strategy_key == 'help':
        if key is None:  # Verifica se não há chave (key)
            Helper.execute()  # Chama a função sem argumentos
        else:
            Helper.execute(key)  # Chama a função com a chave (key)
    elif strategy_key in strategy_map:
        if key is None:  # Verifica se não há chave (key)
            strategy_map[strategy_key]()  # Chama a função sem argumentos
        else:
            strategy_map[strategy_key](key, parametros)  # Chama a função com a chave (key) e parâmetros
    else:
        print(f"comando '{strategy_key}' não encontrado")


# definição de facades
def sysLoader():
    Loader(argv[1:])

def inputLoader():

    
    run = True
    while(run):
        local = path.basename(getcwd())
        entrada = input(f"{local} smr> ")
        entrada = entrada.split(" ")
        if entrada[0] == "exit":
            run = False
            continue
        if len(entrada) >= 1:
            Loader(entrada)
            continue
        error(entrada)
    
    
def error(entrada):
    print(f"comando '{' '.join(map(str, entrada))}' invalido ou faltante, digite 'help' para obter ajuda")

# Verifica se há pelo menos dois argumentos (o primeiro é o nome do script e o segundo é a chave)
if len(argv) > 2: sysLoader()
else: inputLoader()

File: smr.go
package main

import (
	"StudyMonorepo/Command"
	"bufio"
	"fmt"
	"os"
	"strings"
)

var CommandMap = map[string]Command.ICommand{}
var EspecialCommandMap = map[string] func (Command.ICommand){}

func Execute(cmd string, key string, args []string) {
	if command, ok := EspecialCommandMap[cmd]; ok {
		switch cmd {
			case "help":
				if key == "" {
					CommandMap[cmd].Execute("help", &Path, nil)
					return
				}
				if target, ok := CommandMap[key]; ok {
					command(target)
				}
		}
		return
	}
	if command, ok := CommandMap[cmd]; ok {
		command.Execute(key, &Path, args)
		return
	}

	defer fmt.Println("Comando ", cmd, " não encontrado, digite help para ver os comandos disponíveis")
}

func ParseCommand(input string) (string, string, []string) {
	parts := strings.Fields(input)

	switch len(parts) {
	case 0:
		return "", "", nil
	case 1:
		return parts[0], "", nil
	case 2:
		return parts[0], parts[1], nil
	default:
		return parts[0], parts[1], parts[2:]
	}
}

var Path string = "."

func main() {
	CommandMap["create"] = Command.CommandFactory(1)
	CommandMap["delete"] = Command.CommandFactory(2)
	// CommandMap["test"] = Command.CommandFactory(3)
	CommandMap["help"] = Command.CommandFactory(4)
	CommandMap["monorepo"] = Command.CommandFactory(5)
	CommandMap["start"] = Command.CommandFactory(6)
	CommandMap["list"] = Command.CommandFactory(7)

	CommandMap["mono"] = CommandMap["monorepo"]
	CommandMap["run"] = CommandMap["start"]


	EspecialCommandMap["help"] = Command.Help

	for forever := true; forever; {
		scanner := bufio.NewScanner(os.Stdin)

		var input string

		fmt.Print(Path, " smr> ")
		scanner.Scan()
		input = scanner.Text()
		if input == "exit" {
			forever = false
			break
		}
		Execute(ParseCommand(input))
	}

}
